### ISSUE TITLE
Implement prototype `FunctionWrapper` idea

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#1481]
This idea is taken from #1455.  In essence it would be really nice to allow the user to have a way to add some additional behavior to optimizers, like early stopping, or printing the loss, or something like this.  In the discussion there, we threw around a couple of ideas, but here's one I think would be worth implementing.  Eventually I will manage to do this myself, but if someone is looking for some neat thing to work on, I think this issue could be a lot of fun to implement.

Here's the goal---we'd like to be able to write something like this:

```
opt.Optimize(function, parameters, std::make_tuple(EarlyStopping(), PrintLoss()));
```

and then our implementation will look something like this:

```
// syntax is probably not perfect here, I am writing as I go without checking :)
template<typename FunctionType, typename... Targs>
void Optimize(FunctionType& f, arma::mat parameters, std::tuple<Targs...> t)
{
  FunctionWrapper<FunctionType, Targs...> fw(f, t);
}
```

and then FunctionWrapper::Evaluate() is something that calls Evaluate() on the first item in the tuple, which then passes it to the next:

```
class PrintLoss
{
  // I'm almost certain this doesn't work quite right but the idea is there...
  // I guess we have to check for the maximum value of Index.  Maybe it can be simplified.
  template<size_t Index, typename... Targs>
  double Evaluate(const arma::mat& parameters, std::tuple<Targs...>& t)
  {
    const double loss = std::get<Index>(t).Evaluate<Index + 1, Targs...>(parameters, t);
    std::cout << "Hi there, the loss is " << loss << "!" << std::endl;
  }
};
```

For accessing optimizer parameters, etc., I wonder if some combination of a traits class and public accessors could be used.  Possibly, this interface idea of passing some `std::tuple(...)` filled with behaviors could be wrapped for some methods as part of the `Train()` call.

**Comment by scrcro** [comment]
There are a couple of questions:
1. I am not sure how the clients of the optimizers would know what functions can be called from the optimizers? I mean what are some of the interesting events occurring while the optimizers are working? This information is only available to the optimizers. Hence we should let the optimizers decide which functions could be available. So registering some callbacks for certain events in the optimizer seems to be better. 
2. There is an implicit dependency on the order of the calls if we set them up in a tuple. Might be little tricky for the clients, without the knowledge of the internals of the optimizers. Is the data ready for some function to be called?

**Comment by rcurtin** [comment]
Thanks for the response.  It's always good to think everything through first and I think you pointed out some issues that are definitely worth thinking about...

1.  When you say "clients of the optimizers", do you mean the proposed classes like `EarlyStopping` or `PrintLoss`?  For some of those, no optimizer-specific information is necessary.  Others can be restricted to certain optimizers via `static_assert()`s or something like this: imagine a `StopWhenLearningRateIsSmall` class (ok that's an awful name but still) that only works with SGD-like optimizers that have an adaptive step size.  For this case, the optimizer must have a `StepSize()` function, and it's fairly straightforward to check at compile time that it does and issue an error if not.

2. You're right, there is some potential ordering issue here.  But I think that generally this shouldn't produce any serious problems.  Can you think of an example where it would cause a serious issue?  I don't mind mentioning in documentation that there is an order in which each function will be called, but I don't think that it would be a thing users would necessarily need to know about for most common use cases.

Let me know what you think. :)

**Comment by scrcro** [comment]
Thank you for looking at it. 
What I mean by clients of Optimizers are the models or programs that use the optimizers. Not the classes per se. I might be interested when a batch started, or a batch finished or perhaps other some interesting events happening as the optimizer is working. Or what is the progress -- progress bar like ability. I'm kinda influenced by Keras here. 

Again for the next one, I think it would be preferable that the loss is returned after the progress report for a batch. If it happens in between the clients need to handle it. 

I was thinking that each optimizer can have some default function objects (type std::function) which does nothing, and call these when appropriate. Provide additional constructors to pass on the call back function objects. Clients can decide to bind these to their class member functions. This way the optimizers decide when to call and what parameters they can pass and the clients decide what they want to do with the data. Thoughts?

**Comment by zoq** [comment]
> What I mean by clients of Optimizers are the models or programs that use the optimizers. Not the classes per se. I might be interested when a batch started, or a batch finished or perhaps other some interesting events happening as the optimizer is working. Or what is the progress -- progress bar like ability. I'm kinda influenced by Keras here.

I guess nothing hinders us to pass a class instantiation that likes to receive intermediate results, so

```c++
opt.Optimize(function, parameters, std::make_tuple(EarlyStopping(), PrintLoss()));
```

might become

```c++
opt.Optimize(function, parameters, std::make_tuple(EarlyStopping(), PrintLoss(myObject)));
```

> Again for the next one, I think it would be preferable that the loss is returned after the progress report for a batch. If it happens in between the clients need to handle it.

```c++
class PrintLoss
{
  // I'm almost certain this doesn't work quite right but the idea is there...
  // I guess we have to check for the maximum value of Index.  Maybe it can be simplified.
  template<size_t Index, typename... Targs>
  double Evaluate(const arma::mat& parameters, std::tuple<Targs...>& t)
  {
    const double loss = std::get<Index>(t).Evaluate<Index + 1, Targs...>(parameters, t);
    std::cout << "Hi there, the loss is " << loss << "!" << std::endl;
  }
};
```

Will report the loss after a single batch, maybe I missed something?

> I was thinking that each optimizer can have some default function objects (type std::function) which does nothing, and call these when appropriate. Provide additional constructors to pass on the call back function objects. Clients can decide to bind these to their class member functions. This way the optimizers decide when to call and what parameters they can pass and the clients decide what they want to do with the data. Thoughts?


I like the idea to use ``std::function`` the only issue I have is that this could increase the complexity of the optimizer class since it would potentially look like:

```c++
double loss = Evaluate(...);
PrintLossFunction(loss);
CurrentBatch(batchNumber);
...
```

**Comment by rcurtin** [comment]
Agreed with Marcus here.  We can actually see the idea here as a similar template-metaprogram-ey way of accomplishing the same thing as Keras callbacks.  Essentially any of these function wrappers would be called each time the actual function's `Evaluate()` or `Gradient()` functions were called (or `EvaluateWithGradient()` I guess).  It's possible that maybe it would be useful to expand that, but probably the first thing to do is a proof of concept with `PrintLoss()`, and see what we think of it.  Certainly nothing is set in stone, so if we find out that there's some important thing missing from this abstraction we can change it during the review process. :)

(That said, my time is still pretty tight so I have not had a chance to look into any implementation yet.)

**Comment by zoq** [comment]
The ``FunctionWrapper`` class could first run the actual function e.g. ``f.Evaluate()`` and afterwards run all callbacks.

Btw. I did some testing and I'm not sure attaching to ``Evaluate/Gradient/EvaluateWithGradient`` is the best option. I think the current idea is:

```c++
template<...>
void Optimize(FunctionType& f, arma::mat parameters, std::tuple<Targs...> t)
{
  FunctionWrapper<...> fw(f, t);
}

template<...>
class FunctionWrapper
{
	FunctionWrapper(f, t) {};

	Evaluate(...)
	{
		loss = f.Evaluate(...);

		// Invoke callbacks.
		stop = for_each t run Evalute with loss parameter etc. and get the return value to e.g. stop the optimization process.

		return loss;
	}
}
```

so, every time ``Evaluate`` is called we invoke, user-specified callbacks. The optimizer can check if we have to stop the optimization process via some function that is provided by the ``FunctionWrapper``class. Some issues I see with the approach:

- no way to distinguish between immediate ``Evaluation`` calls and actual optimization calls, see SVRG or SAHRA optimizer
- passing optimizer specific parameter is difficult we would have to pass an optimizer instance and provide an interface for each option

The other idea I have is to provide an independent function that is invoked at the end of each iteration. Something like this:

```c++
template<...>
void Optimize(FunctionType& f, arma::mat parameters, std::tuple<Targs...> t)
{
  for (size_t i, ...)
  {
  	// do updates

  	// Invoke callbacks.
	stop = for_each t run Evalute with loss parameter etc. and get the return value to e.g. stop the optimization process.
	if (stop)
		break;
  }
}
```

**Comment by rcurtin** [comment]
Definitely it seems like the original idea has some disadvantages.  But what I really liked about it is that it didn't really require much modification to the optimizer implementations.

What if we instead added more functions that optimizers should call, and then did the same `FunctionWrapper` idea on those?  Like these:

 - `TakeStep(const arma::mat& parameters)`: called when the optimizer took a step to the new matrix `parameters`
 - `EndEpoch(const size_t i, const arma::mat& parameters)`: called only by decomposable function optimizers at the end of an epoch
 - `BeginEpoch(const size_t i, const arma::mat& parameters)`: called only by decomposable function optimizers at the beginning of an epoch
 - `BeginOptimization(const arma::mat& parameters);`
 - `EndOptimization(const arma::mat& parameters);`

Then, the internal optimization code could look like this:

```
template<...>
void Optimize(FunctionType& f, arma::mat parameters, std::tuple<Targs...> t)
{
  FunctionWrapper<FunctionType, Targs...> fw(f, t);

  fw.BeginOptimization(parameters);
  for (size_t i, ...)
  {
    fw.BeginEpoch(i, parameters);
    for (size_t j = 0l j < fw.NumFunctions(); j += batchSize)
    {
      // computation of gradient/objective, etc...

      /// do updates
      parameters += stepSize * gradient; // or whatever the update step is
      fw.TakeStep(parameters);
    }

    fw.EndEpoch(i, parameters);
  }

  fw.EndOptimization(parameters);
}
```

I am not sure how much I like this idea.  The nice thing about it is that an optimizer can expect a unified interface with some number of functions that must be called at various times (and these calls can be optimized out at compile time if no callback uses the results of them), but it seems to me like the wording and naming of functions could be improved.

Actually it is basically exactly the same as @zoq's improved idea, just instead of doing a `for_each t run Evaluate ...`, we just have the optimizer call `FunctionWrapper::<CallbackName>()` (for, e.g., `<CallbackName> = TakeStep`, and then `FunctionWrapper` will handle looping over the different callback objects that were passed in the tuple.  Putting all that in `FunctionWrapper` is probably a good idea so that we reduce duplication of code in the different optimizers.

In any case, no matter what it is we do, we will have to at least establish the points at which callbacks can be used (epoch begin, epoch end, etc.).

**Comment by scrcro** [comment]
> In any case, no matter what it is we do, we will have to at least establish the points at which callbacks can be used (epoch begin, epoch end, etc.).

Precisely. The optimizer must decide what to call when. This solution appears to be much better. 
I will give it a shot.

**Comment by scrcro** [comment]
> Putting all that in `FunctionWrapper` is probably a good idea so that we reduce duplication of code in the different optimizers.

Not sure if this is do-able without touching the Optimizer code. I mean it would e required to call atleast the `fw.Callback`. I think changing the optimizer interface is a simpler solution. Just have a set of `noop` functions. Allow callers of the optimizer to set these callbacks that they are interested in. Thoughts?

**Comment by rcurtin** [comment]
Oh, definitely the optimizers would have to be updated.  I just mean that we could update the optimizers only to include some call like `fw.Callback`, and then inside the `FunctionWrapper` class we can write the actual code that traverses through the tuple of callback objects and calls their corresponding `Callback` functions (if they exist).

**Comment by zoq** [comment]
@scrcro I could provide some example code that could be helpful.

**Comment by zoq** [comment]
@scrcro https://github.com/mlpack/ensmallen/pull/49 is an initial idea, I think since this is a PR, commenting on the code is easier.

**Comment by zoq** [comment]
I think that covers all points. I'm not sure if we need the FunctionWrapper class as Ryan suggested or if a static method is sufficient. I guess we if decide to provide different callbacks like ``BeginOptimization`` we should also think about the function names.

**Comment by scrcro** [comment]
Right. Even the function signatures needs to be determined as well. The wrapper is only required as a convenience for the caller. They do not have to necessarily provide all the methods if they are not interested.

**Comment by rcurtin** [comment]
Let's continue the discussion in ensmallen#49.  I'll close this issue. :+1:
